swarm/swap/swap.go

```diff
@@ -0,0 +1,589 @@
+}
+
+type EntryDirection bool
+
+const (
+	DebitEntry  EntryDirection = true
+	CreditEntry EntryDirection = false
+)
+
+type SwapAccountedMsgType interface {
+	GetMsgPrice() *big.Int
+}
+
+//Handler for received messages
+func (sp *SwapPeer) RunAccountedProtocol(protocolHandler func(ctx context.Context, msg interface{}) error) error {
+	//the `peer.Run` function is a loop, so in order to pre-/post-process a message with accounting,
+	//we need to save the actual handler
+	sp.handlerFunc = protocolHandler
+
+	//then run the handler loop function
+	return sp.Run(sp.handleAccountedMsg)
+}
+
+//get a peer's balance
+func (swap *Swap) GetPeerBalance(peer discover.NodeID) *big.Int {
+	if p, ok := swap.peers[peer]; ok {
+		return p.balance
+	}
+	return nil
+}
+
+//Handle a received message; this is the handler loop function.
+//Check if it needs accounting, and if yes, apply accounting logic:
+//Check for sufficient funds, perform operation, then account
+func (sp *SwapPeer) handleAccountedMsg(ctx context.Context, msg interface{}) error {
+	var err error
+	var price *big.Int
+
+	//the message is one which needs accounting...
+	if _, ok := msg.(SwapAccountedMsgType); ok {
+		//..so first check if there are enough funds for the operation available
+		//(for crediting, this means if we are not essentially "overdrafting", or crossing the threshold)
+		price, err = sp.checkAvailableFunds(ctx, msg, CreditEntry)
+		//if not (or some other error occured), return error
+		if err != nil {
+			//also, if the error is indeed insufficient funds, then disconnect the peer
+			if err == ErrInsufficientFunds {
+				log.Error("Insufficient funds, dropping peer")
+				sp.Drop(err)
+			}
+			return err
+		}
+		//at this point we know there are sufficient funds, so process the message
+		err = sp.handlerFunc(ctx, msg)
+		if err == nil {
+			//and if no errors occurred, finally book the entry
+			sp.AccountMsgForPeer(ctx, msg, price, CreditEntry)
+		}
+	} else {
+		//this message doesn't need accounting, so just process it
+		err = sp.handlerFunc(ctx, msg)
+	}
+	return err
+}
+
+//Send a message
+//Check if it needs accounting, and if yes, apply accounting logic:
+//Check for sufficient funds, perform operation, then account
+func (sp *SwapPeer) Send(ctx context.Context, msg interface{}) error {
+	var err error
+	var price *big.Int
+
+	//the message is one which needs accounting...
+	if _, ok := msg.(SwapAccountedMsgType); ok {
+		//..so first check if there are enough funds for the operation available
+		price, err = sp.checkAvailableFunds(ctx, msg, DebitEntry)
+		//if not (or some other error occured), return error
+		if err != nil {
+			//also, if the error is indeed insufficient funds, then disconnect the peer
+			if err == ErrInsufficientFunds {
+				log.Error("Insufficient funds, dropping peer")
+				sp.Drop(err)
+			}
+			return err
+		}
+		//at this point we know there are sufficient funds, so process the message
+		err = sp.Peer.Send(ctx, msg)
+		if err == nil {
```

Comment:
here I see why you want 2 calls, 
- if peer has no credit the send should not happen
- sending fails you dont want to account for the peer either way

but i am convinced the logic can be simplified, 
you just send the message,
-  if that brings you into red, which cannot happen normally since you top up your balance and if that fails you fail/quit earlier or you wanna play with fire, and then you can just try your luck sending  the message 
- if this message brings your peer into red, its fine this is how we define the threshold, 

so i suggest just do a send and account if successful

Reply:
I am not sure we are taking the same basis to define *threshold*.

> if that brings you into red, which cannot happen normally since you top up your balance and if that fails you fail/quit earlier...

sounds to me you are coming from the basis that we have to top-up, *starting from zero*. You can't get into red basically.

I am basing my self on the same paper above, Figure 2: Swap balance and swap thresholds. There we have a balance floating between a negative value and a positive value. So we need to be able to go into red until the thresholds are reached.

We have 2 thresholds:
* The threshold when a peer needs to trigger a cheque payment ("Payment threshold" in Figure 2 of the paper). If a peer gets into red and crosses that threshold, I understand this to be just fine. We can continue accounting, but the peer should issue a cheque

* The threshold at which the peer would get disconnected ("Disconnect threshold" in the paper). If a peer gets into red and crosses this one, then one would expect that to be the "final, uncrossable" line - you can't get over that. If you do, you get disconnected. 

Now we need to look at the details of this: Let's say `disconnect threshold` is -100. I am at -98, and I want to send a message with price 3. We have 3 options:
* We first send, then account. Sending will succeed, then we account, we detect balance at -101 , the accounting returns a  "insufficient funds* error which is misleading because the message send succeeded. The remote peer just just drops us. In the case of sending a `RetrieveRequest` (the only accounted message so far), what's the result? Probably the retrieval fails as the remote peer just dropped us. 

* In the previous option we actually have to consider that we don't really know the message price. What if the price is 100? The we go from -98 to -198, the same applies as above but we over-drafted (got into red) quite considerably. This may be insignificant for small amounts, but we don't have the full extent of accounting defined yet so it is a **risk**.

* We first check for available funds, we see the balance at -98 and return immediately a "insufficient funds" error. This is how accounting usually works.